### DISCUSSION TITLE
Yamaha KX-49/KX-25 keyboards are recognized, but no data in the midi debug monitor

### DISCUSSION THREAD

**Discussion by ghost** [probonopd/MiniDexed#176] [Q&A] [answered]
I have been building and trying MiniDexed the last 6 days on Raspberry Pi Zero, Raspberry Pi 1B and Raspberry Pi 3B, but on neither of them i got sound out of MiniDexed when connecting my Yamaha KXs ( https://synthpedia.net/yamaha/kx49/ ) via USB. First i thought i messed up the configuration files or its sending on the wrong midi channel, but after trying out all different kind of combinations and switching on MIDIDumpEnabled=1, i discovered that other midi controller from my music gear arsenal are sending data which is received by minidexed, but not the KX keyboards.

I tried :
- Fishman Triple Play Guitar Midi Controller which is working fine, even ProgrammChanges
- Presonus Atom, a drumpad controller like Maschine...which almost fried the Raspberry as the controller was sucking so much power that i pulled the usb plug after some seconds....but MiniDexed received midi data and was producing sounds.

Like the other two devices, the Yamaha KX keyboards are recognized and configured as you can see on the screenshots...but no data when being played.
![YamahaKX](https://user-images.githubusercontent.com/38320327/165161722-176132e2-00a8-4ce0-bc60-79bda60df89e.jpg)
![PresonusAtom](https://user-images.githubusercontent.com/38320327/165161841-199d3aa4-5fc6-447f-b7df-76852b5b0387.jpg)
![Bildschirmfoto vom 2022-04-25 21-15-38](https://user-images.githubusercontent.com/38320327/165161930-49d60df2-d5e2-483f-9418-6d7998bb2557.png)

Now heres the thing: The KX keyboards are presenting 3 midi ports when used in windows, MacOS or even in Linux which are YAMAHA KX MIDI 1, .. MIDI 2 and ...MIDI 3 (screenshot is from Dexed runing on ArchLinux, working perfectly fine there).
- MIDI 1 is for all the keys and the 4 encoder, which are doubled...so 8 of them, sending CCs
- MIDI 2 is for the transport section of the keyboard, which is for DAW usage only...MCU
- MIDI 3 is exclusively for a external Yamaha patch templates exchange software...not used anymore

From the Yamaha KX Manual : 
MIDI ports on the KX
For data transfer via a USB cable between the
KX and the computer, you can use three MIDI
ports by installing the USB-MIDI driver. Each of
three MIDI ports can handle the different MIDI
messages as described below.
- MIDI Port 1:
This port is indicated as “Yamaha KX-1” on
Windows and “YAMAHA KX Port 1” on
Macintosh. This port is used for data transfer of
the musical performance such as keyboard
performance (note on/off), Modulation wheel
operation (Control Change) and Pitch Bend
wheel operation.
- MIDI Port 2:
This port is indicated as “Yamaha KX-2” on
Windows and “YAMAHA KX Port 2” on
Macintosh. This port is used for data transfer
generated by using the DAW REMOTE buttons
(with the exception of the Cursor buttons, [ESC]
button and [ENTER] button) and CONTROL
knobs.
- MIDI Port 3:
This port is indicated as “Yamaha KX-3” on
Windows and “YAMAHA KX Port 3” on
Macintosh. This port is used for data transfer
generated by operating the KX Editor on the
computer.

My guess is that MiniDexed is messing up the 3 MidiPorts when doing the configuration and is listening on the wrong port , maybe Port 3 where theren aren´t any midi notes send out from playing the keyboard ??

"I am owner of a Yamaha KX-88 (the old tank from 1985 https://yamahablackboxes.com/collection/yamaha-kx88-controller/ ) for 12 years now, but sadly the master midi keyboard collected some dust over the last years. Would love to use it the way it was built back then in combination with TX816 modules where the A, B and A + B switching with different channel sounds really shines. As soon as i will get the serial midi port working i can report some infos regarding channel aftertouch, as the KX88 is able to do so. For now i am still trying to get one of my cheap chinese plastic housing built Yamaha KX (KX49 and KX25) from the 2000er years recognized and working. For some strange reason there is no midi data in debug mode showing up, and no sound because of that...."

_Originally posted by @StephanBeer in https://github.com/probonopd/MiniDexed/discussions/162#discussioncomment-2630192

_

**Comment by eLuSiVeMiTe** [comment]
Is it class compliant.  (Will it run on windows with no drivers) 
Idle curiosity.


Have you tried running midi din in via a usb interface?

**Comment by probonopd** [comment]
Hello @StephanBeer, please remove `usbspeed=full` from the file `cmdline.txt` on the SD card and try again.

**Comment by ghost** [comment]
Hey @probonopd and @eLuSiVeMiTe, first off thanks for your work on that wonderfull and exciting project !!

I removed the usbspeed line so that cmdline.txt is blank, but that did not change anything, there is still no midi data incomeing and received from the KX keyboard. Messages from the Triple Play are still received and show up on the screen, PCs and Volume via Fishman Controller working fine.

When putting the same sdcard from the Pi Zero into the Pi 1B, it seems that the Debug Monitor is frozen and no messages are showing up, like when unplug or replug the controller. Is the problem for the Raspberry Pi 1 resolved ?? Thats with the newest build...

I would love to have a "confirmed working with following USB devices" section in the WIKI someday, which would be helpful to exclude possible errors when building the project. Right now its often hit and miss for me, so its hard to do a systematic debugging to find the reason when things won´t work.

**Comment by luisfcorreia** [comment]
@StephanBeer I'll ask again, is that device MIDI class compliant?

**Comment by eLuSiVeMiTe** [comment]
> Hey @probonopd and @eLuSiVeMiTe, first off thanks for your work on that wonderfull and exciting project !!


I'm just along for the ride. Wouldn't know where to start with programming. (But I could build you a house) 

I had a good dig last night and I can't find any mention of that controller being class compliant (runs on standard drivers) 
Im assuming midi input devices need to be class compliant to work on the pi 

Try using a usb midi interface and midi din from the Yamaha.

**Comment by ghost** [comment]
> @StephanBeer I'll ask again, is that device MIDI class compliant?

I have been using the Yamaha KX USB keyboards for about 14 years now, on Windows XP, Windows 7, 8 and now 10. As i switched to Linux (for music, recording and composing Windows is stil needed) the last 2 years, the Yamaha KX are gladly showing up in every application or with pipewire, but i guess thats because they are present in the Linux Kernel ? To cut it short i never have tried to use them without a driver. I have to remove the drivers and plug in one of the keyboards to see if they are working without, or even show up in the device manager.... are there other ways to find out USB class compliance ? Plug them into ones android phone ?

**Comment by luisfcorreia** [comment]
plug them to an iPad or any Android with midi enabled OTG

**Comment by ghost** [comment]
> 

Plugged the keyboard via USB type B to micro USB to USB C adapter into my android phone, and it is enabled only as a physical keyboard....a keyboard on which you type on, not a midi keyboard. That must be because of the Midi port 2, which does Computer keyboard actions. Normally there should be a dialog popup where you can change to midi device ?

**Comment by ghost** [comment, marked as answer]
> 



> > Hey @probonopd and @eLuSiVeMiTe, first off thanks for your work on that wonderfull and exciting project !!
> 
> I'm just along for the ride. Wouldn't know where to start with programming. (But I could build you a house)
> 
> I had a good dig last night and I can't find any mention of that controller being class compliant (runs on standard drivers) Im assuming midi input devices need to be class compliant to work on the pi
> 
> Try using a usb midi interface and midi din from the Yamaha.

It really seems that its not Class compliant, so i will have to switch the Output Mode of the keyboard from USB to the traditional MIDI ports and finish building the serial MIDI ports on my MiniDexed build.

**Comment by pik33** [comment]
I have a compatibility problem with Novation Impulse 61. They say "class compliant" but MiniDexed doesn't like it. Now I found this thread. Of course, adding a standard MIDI 5-pin input is a workaround, but not a solution. A synth type program/device should allow to recognize and use popular MIDI controllers. 

Maybe the problem is caused by these multi-port interfaces where only one port/device is a MIDI device while the rest of them are something else making Circle/MiniDexed not working.

So the next step for me is to connect the Impulse to a Linux box and check what devices are listed by lsusb.

**Reply by probonopd** [reply]
> Maybe the problem is caused by these multi-port interfaces where only one port/device is a MIDI device while the rest of them are something else

Possibly. It does work with the nectar Imact LX61+, which also has 2 MIDI devices, though.

**Reply by diyelectromusic** [reply]
> A synth type program/device should allow to recognize and use popular MIDI controllers.

A synth type /program/ would typically be used on a fully bloated commercial OS that handles a full USB device stack and it is the manufacturers problem to ensure their devices will work on that stack.  There aren't many (any?) synth type /devices/ that would act as a USB host and those that do will also probably use an OS.

MiniDexed is an extremely cut-down environment where everything is done "by hand" in the code so supporting cases where a manufacturer has not gone with full class compliance is a lot of work, no matter how popular the device is.

If you want a "will support anything" version of Dexed, then I would recommend running a software synth version on a PC; investigate the version of Dexed on Zynthian on a Raspberry Pi; or maybe see if you have any better luck with microDexed on the Teensy.

Of course, you could try to petition the manufacturer of your device to release a full MIDI class compliant version :)

I'm very happy to keep following your experiments, looking at your device and trying to work out if there is a bug in the class compliant support in MiniDexed or if this is a simple edge case that can be accommodated, but it isn't easy not having the actual device myself :)

**Reply by diyelectromusic** [reply]
> I'm very happy to keep following your experiments, looking at your device and trying to work out if there is a bug in the class compliant support in MiniDexed or if this is a simple edge case that can be accommodated, but it isn't easy not having the actual device myself :)

I might have made some progress.  I'm not sure if it's a bug, but more "not supported at present".  It looks like any class compliant devices using USB INTERRUPT endpoints (of which the Novation devices seem to be) are not supported in the Circle environment.  It only looks for BULK endpoints.  Just hacking the code to look does seem to allow them to work, but not being a USB person, I don't know the implications of that.  But if you want to follow how this develops, I've logged it as an issue here: https://github.com/rsta2/circle/issues/316.

I don't know if this is relevant to the Yamaha discussion though...

Kevin

**Reply by probonopd** [reply]
> I have a compatibility problem with Novation Impulse 61

Thanks for confirming over at https://github.com/rsta2/circle/issues/316?#issuecomment-1257158364 that the Novation Impulse 61 is working now. And thanks @diyelectromusic and @rsta2 for making it work :+1:

**Comment by diyelectromusic** [comment]
I wonder if it is worth trying the usbignore= option in cmdline.txt to tell MiniDexed to ignore the non-MIDI USB interfaces being presented?  Basically anything that isn't int1-3-0 maybe?  See https://github.com/probonopd/MiniDexed/wiki/Files#cmdlinetxt
